### PR TITLE
feat(audio): add ACE-Step 1.5 music generation support

### PR DESCRIPTION
**Pull request by Minamiyama**
## Summary

Add built-in ACE-Step 1.5 support for text-to-music generation.

## Changes

- Register `ACE-Step1.5` as an audio model using the PyTorch engine.
- Support the complete checkpoint bundle from:
  - Hugging Face `ACE-Step/Ace-Step1.5`, revision `main`
  - ModelScope `ACE-Step/Ace-Step1.5`, revision `master`
- Use the bundled `acestep-v15-turbo` DiT and official VAE.
- Support the optional bundled `acestep-5Hz-lm-1.7B` with `pt`, `vllm`, and `mlx` backends.
- Expose music generation through the existing Speech API:
  - lyrics through `input`
  - music description through `instruct`
  - configurable duration, seed, BPM, key, time signature, inference steps, guidance, and LM reasoning options
  - AAC, FLAC, MP3, Opus, WAV, and WAV32 output
- Add Hugging Face and ModelScope download-source metadata.
- Add per-model virtual environment dependencies and Python 3.11/3.12 validation.
- Isolate mutable model-side Python files in a temporary checkpoint workspace while reusing immutable model weights.
- Add launch, API usage, parameter, format, and hardware documentation.

## Vendored runtime

Vendor the inference-only ACE-Step 1.5 runtime from upstream commit:

`14c0211d5a0653b0f63e27686f4c3f151b4d8629`

The vendored source includes the required ACE-Step inference components and
bundled `nano-vllm` runtime. Training code, datasets, API servers, CLI/Gradio
UI, tests, examples, development metadata, and unused DiT variants are
excluded.

Runtime dependencies such as PyTorch, Transformers, Diffusers, and TorchAO
remain managed by the Xinference model virtual environment. Model launch no
longer requires installing ACE-Step source directly from GitHub.

The upstream MIT licenses and source revision information are retained.

## Current scope

This initial integration supports the components included in the complete
ACE-Step 1.5 checkpoint bundle:

- `acestep-v15-turbo`
- official VAE
- optional `acestep-5Hz-lm-1.7B`

Arbitrary standalone DiT, LM, and VAE combinations are left for follow-up work.

### CODE REVIEW

File: xinference/model/audio/ace_step.py [hunk 1]
```diff
@@ -0,0 +1,460 @@
...
+                model_path,
+                checkpoint_dir,
+                copy_function=self._link_checkpoint_file,
+            )
+        except BaseException:
+            workspace.cleanup()
+            raise
+        return workspace, checkpoint_dir
+
+    def load(self) -> None:
```
Comment (qinxuye) [inline xinference/model/audio/ace_step.py:211]: Please add focused regression tests for this new adapter. At minimum, cover load-time option validation and initialization-failure cleanup, checkpoint workspace isolation, and the `speech()` validation plus success/failure output paths. The existing audio-engine tests only exercise registration and never execute `AceStepModel.load()` or `AceStepModel.speech()`, so the current green CI would not catch regressions in this integration.

File: xinference/model/audio/tests/test_ace_step.py [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+        runtime_config.write_text("ORIGINAL = False")
+        assert source_config.read_text() == "ORIGINAL = True"
+    finally:
+        workspace.cleanup()
+
+    assert not workspace_path.exists()
+
+
+@pytest.mark.parametrize(
+    ("request", "message"),
```
Comment (qinxuye) [inline xinference/model/audio/tests/test_ace_step.py]: [P1] Rename this parametrized argument from `request`. Pytest reserves that name for its built-in fixture and aborts collection with `request is a reserved name and cannot be used in @pytest.mark.parametrize`, so none of the new adapter tests run. Please use a name such as `speech_kwargs` and rerun this test file. Also run pre-commit: mypy currently reports four `ModuleType` attribute errors in the fake runtime helper above.

File: xinference/thirdparty/ace_step_1_5/acestep/core/generation/handler/generate_music_decode.py [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+                if not using_mlx_vae:
+                    vae_cpu = os.environ.get("ACESTEP_VAE_ON_CPU", "0").lower() in ("1", "true", "yes")
+                    if not vae_cpu:
+                        if self.device == "mps":
+                            logger.info(
+                                "[generate_music] MPS device: skipping VRAM check "
+                                "(unified memory), keeping VAE on MPS"
+                            )
+                        else:
+                            effective_free = get_effective_free_vram_gb()
```
[the remark this answers is not in the document]
Comment (Minamiyama) [reply xinference/thirdparty/ace_step_1_5/acestep/core/generation/handler/generate_music_decode.py]: Addressed in 39110491: the VAE decode VRAM check now queries `_cuda_device_index(self.device)` instead of implicitly using GPU 0.

File: xinference/thirdparty/ace_step_1_5/acestep/core/generation/handler/init_service_orchestrator.py [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+            if config_path is None:
+                config_path = "acestep-v15-turbo"
+                logger.warning(
+                    "[initialize_service] config_path not set; defaulting to 'acestep-v15-turbo'."
+                )
+
+            resolved_device = self._resolve_initialize_device(device)
+            self.device = resolved_device
+            self.offload_to_cpu = offload_to_cpu
+            self.offload_dit_to_cpu = offload_dit_to_cpu
```
[the remark this answers is not in the document]
Comment (Minamiyama) [reply xinference/thirdparty/ace_step_1_5/acestep/core/generation/handler/init_service_orchestrator.py:81]: Addressed in 39110491: initialization now calls `torch.cuda.set_device(...)` for any resolved CUDA device before subsequent runtime setup. The same check also handles dtype selection for `cuda:N`.

File: xinference/thirdparty/ace_step_1_5/acestep/core/generation/handler/init_service_memory_basic.py [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+                target_type = torch.device(str(target_device)).type
+        except Exception:
+            target_type = str(target_device).strip().lower().split(":", 1)[0]
+            if not target_type:
+                logger.warning(
+                    "[_is_on_target_device] Malformed target device value: {!r}",
+                    target_device,
+                )
+                return False
+        return tensor.device.type == target_type
```
[the remark this answers is not in the document]
Comment (Minamiyama) [reply xinference/thirdparty/ace_step_1_5/acestep/core/generation/handler/init_service_memory_basic.py]: Addressed in 39110491: `_is_on_target_device` now compares both device type and explicit index, and resolves bare `cuda` against `torch.cuda.current_device()`.

File: xinference/thirdparty/ace_step_1_5/acestep/core/generation/handler/generate_music.py [hunk 6]
```diff
@@ -0,0 +1,506 @@
...
+        if guidance_scale > 1.0:
+            dit_key = "xl_base" if is_xl else "base"
+        else:
+            dit_key = "xl_turbo" if is_xl else "turbo"
+        per_batch_gb = DIT_INFERENCE_VRAM_PER_BATCH.get(dit_key, 0.6)
+        # Longer audio = more latent frames (5 Hz rate) = more memory.
+        duration_factor = max(1.0, duration_s / 60.0)
+        needed_gb = per_batch_gb * actual_batch_size * duration_factor + VRAM_SAFETY_MARGIN_GB
+
+        free_gb = get_effective_free_vram_gb()
```
[the remark this answers is not in the document]
Comment (Minamiyama) [reply xinference/thirdparty/ace_step_1_5/acestep/core/generation/handler/generate_music.py]: Addressed in 39110491: the VRAM preflight now passes `_cuda_device_index(self.device)`. Bare `cuda` resolves to the active CUDA device; explicit `cuda:N` keeps its index.

File: xinference/thirdparty/ace_step_1_5/acestep/constants.py [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+
+# Supported languages for vocal generation and language detection
+# Covers major world languages with good TTS support in the underlying model
+# 'unknown' is used when language cannot be determined automatically
+VALID_LANGUAGES = [
+    'ar', 'az', 'bg', 'bn', 'ca', 'cs', 'da', 'de', 'el', 'en',
+    'es', 'fa', 'fi', 'fr', 'he', 'hi', 'hr', 'ht', 'hu', 'id',
+    'is', 'it', 'ja', 'ko', 'la', 'lt', 'ms', 'ne', 'nl', 'no',
+    'pa', 'pl', 'pt', 'ro', 'ru', 'sa', 'sk', 'sr', 'sv', 'sw',
+    'ta', 'te', 'th', 'tl', 'tr', 'uk', 'ur', 'vi', 'yue', 'zh',
```
[the remark this answers is not in the document]
Comment (Minamiyama) [reply xinference/thirdparty/ace_step_1_5/acestep/constants.py:18]: `te` is already present in `[tool.codespell].ignore-words-list` in `pyproject.toml` on this PR branch. The vendored `thirdparty` tree is also excluded from codespell; focused pre-commit verification passes.

File: xinference/thirdparty/ace_step_1_5/acestep/core/generation/handler/conditioning_text.py [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+    ) -> Tuple[List[str], torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor, Optional[torch.Tensor], Optional[torch.Tensor]]:
+        """Tokenize caption/lyric prompts and optional non-cover branch prompts."""
+        actual_captions, actual_languages = self._extract_caption_and_language(parsed_metas, captions, vocal_languages)
+
+        # Detect is_lego_sft from the loaded model config (set on the SFT-stems checkpoint).
+        is_lego_sft = (
+            hasattr(self, "model")
+            and self.model is not None
+            and getattr(self.model.config, "is_lego_sft", False)
+        )
```
[the remark this answers is not in the document]
Comment (Minamiyama) [reply xinference/thirdparty/ace_step_1_5/acestep/core/generation/handler/conditioning_text.py:77]: Addressed in 39110491 with nested `getattr`, so models without a `config` attribute safely use the default.

File: xinference/thirdparty/ace_step_1_5/acestep/core/generation/handler/memory_utils.py [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+    def _get_auto_decode_chunk_size(self) -> int:
+        """Choose a conservative VAE decode chunk size based on available memory."""
+        override = os.environ.get("ACESTEP_VAE_DECODE_CHUNK_SIZE")
+        if override:
+            try:
+                value = int(override)
+                if value > 0:
+                    return value
+            except ValueError:
+                pass
```
[the remark this answers is not in the document]
Comment (Minamiyama) [reply xinference/thirdparty/ace_step_1_5/acestep/core/generation/handler/memory_utils.py:98]: Addressed in 39110491: invalid or non-positive `ACESTEP_VAE_DECODE_CHUNK_SIZE` values now emit an operator-visible warning before automatic sizing.

File: xinference/thirdparty/ace_step_1_5/acestep/audio_utils.py [hunk 10]
```diff
@@ -0,0 +1,622 @@
...
+
+    lora_service = getattr(dit_handler, "_lora_service", None)
+    if lora_service is None or not lora_service.registry:
+        return ""
+
+    hash_obj = hashlib.sha256()
+    found_any = False
+
+    for adapter_name in sorted(lora_service.registry.keys()):
+        meta = lora_service.registry[adapter_name]
```
[the remark this answers is not in the document]
Comment (Minamiyama) [reply xinference/thirdparty/ace_step_1_5/acestep/audio_utils.py]: Addressed in 39110491 with the suggested defensive fallback: `registry.get(adapter_name) or {}`.

File: xinference/model/audio/tests/test_ace_step.py (hunk 2 above)
Comment (Minamiyama) [reply xinference/model/audio/tests/test_ace_step.py]: Addressed in 7467460fc: renamed the parametrized argument to `speech_kwargs` and used `setattr` for dynamic `ModuleType` attributes. `test_ace_step.py` passes all 23 tests, and the branch-wide modified-file pre-commit checks pass.

File: xinference/model/audio/ace_step.py (hunk 1 above)
Comment (Minamiyama) [reply xinference/model/audio/ace_step.py:211]: Addressed in 61f1c7e31: added focused ACE-Step adapter regression tests covering load-option validation, initialization-failure cleanup, checkpoint workspace isolation, and `speech()` validation plus success/failure output paths. The focused file now passes 23 tests.

**Review by qinxuye — APPROVED** [review]
LGTM
